opr/grading.go

```diff
@@ -49,79 +52,67 @@ func CalculateGrade(avg [20]float64, opr *OraclePriceRecord) float64 {
-			}
-		}
+		v.Difficulty = v.ComputeDifficulty(v.Entry.ExtIDs[0])
 	}
+
+	// Throw away all the entries but the top 50 on pure difficulty alone.
+	// Note that we are sorting in descending order.
+	sort.Slice(list, func(i, j int) bool { return list[i].Difficulty > list[j].Difficulty })
+
 	if len(list) > 50 {
-		last = 50
+		list = list[:50]
 	}
-	// Go through and throw away entries that are outside the average or on a tie, have the worst difficulty
-	// until we are only left with 10 entries to reward
-	for i := last; i >= 10; i-- {
+	for i := len(list); i >= 10; i-- {
 		avg := Avg(list[:i])
 		for j := 0; j < i; j++ {
 			CalculateGrade(avg, list[j])
 		}
-		// bubble sort the worst grade to the end of the list. Note that this is nearly sorted data, so
-		// a bubble sort with a short circuit is pretty darn good sort.
-		for j := 0; j < i-1; j++ {
-			cont := false                // If we can get through a pass with no swaps, we are done.
-			for k := 0; k < i-j-1; k++ { // yes, yes I know we can get 2 or 3 x better speed playing with indexes
-				if list[k].Grade > list[k+1].Grade { // bit it is tricky.  This is good enough.
-					list[k], list[k+1] = list[k+1], list[k] // sort first by the grade.
-					cont = true                             // any swap means we continue to loop
-				} else if list[k].Grade == list[k+1].Grade { // break ties with PoW.  Where data is being shared
-					if list[k].Difficulty < list[k+1].Difficulty { // we will have ties.
-						//list[k], list[k+1] = list[k+1], list[k]
-						cont = true // any swap means we continue to loop
-					}
-				}
-			}
-			if !cont { // If we made a pass without any swaps, we are done.
-				break
-			}
-		}
+		// Because this process can scramble the sorted fields, we have to resort with each pass.
+		sort.Slice(list[:i], func(i, j int) bool { return list[i].Difficulty > list[j].Difficulty })
+		sort.SliceStable(list[:i], func(i, j int) bool { return list[i].Grade > list[j].Grade })
 	}
 	tobepaid = append(tobepaid, list[:10]...)
+
 	return tobepaid, list
 }
 
+// RemoveDuplicateMiningIDs runs a two-pass filter on the list to remove any duplicate entries.
+// The entry with higher difficulty is kept.
+// Two passes are used to avoid slice deletion logic
+func RemoveDuplicateMiningIDs(list []*OraclePriceRecord) []*OraclePriceRecord {
+	// miner id => slice index of highest difficulty entry
+	highest := make(map[string]int)
+
+	for i, v := range list {
+		id := fmt.Sprintf("%d+%s", len(v.FactomDigitalID), strings.Join(v.FactomDigitalID, "-"))
+
+		if dupe, ok := highest[id]; ok {
+			if v.Difficulty > list[dupe].Difficulty { // equal difficulties = first come first serve
+				highest[id] = i
+			}
+		} else {
+			highest[id] = i
+		}
+	}
+
+	nlist := make([]*OraclePriceRecord, 0)
+	for i, v := range list {
+		id := fmt.Sprintf("%d+%s", len(v.FactomDigitalID), strings.Join(v.FactomDigitalID, "-"))
+		if valid, ok := highest[id]; ok && i == valid {
```

Comment:
This condition seems suspicious, is it ever possible that the condition is not true? It looks like if it does, then we have a serious problem in our algorithm and we should crash rather than pass it silently.

Another suggestion: why not iterate through the `highest` map instead to build that `nlist`?

Reply:
> This condition seems suspicious, is it ever possible that the condition is not true? It looks like if it does, then we have a serious problem in our algorithm and we should crash rather than pass silently?

it's not possible for `ok` to be false (there's 1+ entry for every id in the list), just a habit of including it in a map check. i suppose it could be removed for readability. `i == valid` is false when there's an entry with the same id but higher difficulty, that's the one to keep

> Another suggestion: why not iterate directly through the highest map instead to build that nlist?
golang maps are iterated through randomly unfortunately. we need to preserve the order

Reply:
yep got it, thanks. I still need to practice my golang reading, still not completely familiar.